pdfminer/pdfdocument.py

```diff
@@ -895,6 +901,105 @@ def search(entry: object, level: int
             return
         return search(self.catalog['Outlines'], 0)
 
+    def get_page_labels(
+        self,
+        total_pages: Optional[int] = None
+    ) -> Iterator[str]:
+        """
+        Generate page label strings for the PDF document.
+
+        If the document includes page labels, generates strings, one per page.
+        If not, raises PDFNoPageLabels.
+
+        If total_pages is specified, it is used to determine when the iteration
+        stops; if not, the iteration is unbounded.
+        """
+        assert self.catalog is not None
+
+        try:
+            labels_tree = dict_value(self.catalog['PageLabels'])
+        except (PDFTypeError, KeyError):
+            raise PDFNoPageLabels
+
+        def walk_number_tree(
+            td: Dict[Any, Any]
+        ) -> Iterator[Tuple[int, Dict[Any, Any]]]:
+            """
+            Walk number tree node dictionary yielding (page index, dict) pairs.
+
+            See PDF spec, section 3.8.5.
+            """
+            if 'Nums' in td:  # Leaf node
+                objs = list_value(td['Nums'])
+                for (k, v) in choplist(2, objs):
+                    yield int_value(k), dict_value(v)
+
+            if 'Kids' in td:  # Intermediate node
+                for child_ref in list_value(td['Kids']):
+                    yield from walk_number_tree(dict_value(child_ref))
```

Comment:
Can you extract a `NumberTree` class in a new *data_structures.py* file that does the parsing and "walking over" logic?

Reply:
I will put some work in extracting a `NumberTree` class now.

Reply:
Thanks!

Reply:
I've refactored this into a `NumberTree` and `PageLabels` class. Let me know what you think.

Reply:
Looks good, thanks. I fixed a few more issues and also placated mypy. You should take another look.